# PokéData: A Data-Driven Victory

__Context:__ Professor Oak's hypothesis is that pokémon have been getting stronger with each generation however some of his colleagues seem to disagree. With the pokémon championship battle coming closer, he wants to use this event to test his theory. He tasks a promising young trainer to fight the undefeated champion Cynthia with a carefully constructed team of 6 pokémon and bring an end to her win streak. 

__Objective:__ Create a team of 6 non-legendary pokémon that can defeat Cynthia's team and analyse trends in the data to in order to assess professor Oak's hypothesis.

## Setup: Cleaning the Data

In [137]:
# import libraries

import pandas as pd
import sqlite3

In [138]:
# loading dataset

df = pd.read_csv('C:/Users/danny/Documents/Projects/PokeData/MainStats.csv')
df.tail()

ID          Name  Total   HP  Attack  Defense  SpAtk  SpDef  Speed  \
1204  1021   Raging Bolt    590  125      73       91    137     89     75   
1205  1022  Iron Boulder    590   90     120       80     68    108    124   
1206  1023    Iron Crown    590   90      72      100    122    108     98   
1207  1024     Terapagos    450   90      65       85     65     85     60   
1208  1025     Pecharunt    600   88      88      160     88     88     88   

         Type1    Type2  Height  Weight  
1204  Electric   Dragon     NaN     NaN  
1205      Rock  Psychic     NaN     NaN  
1206     Steel  Psychic     NaN     NaN  
1207    Normal      NaN     NaN     NaN  
1208    Poison    Ghost     NaN     NaN

This data is missing some key information that we will need such as the generation of each Pokémon and their legendary status. To fix this, we need to find the relevant data and append it to the above dataset.

In [139]:
# loading other dataset

df2 = pd.read_csv('C:/Users/danny/Documents/Projects/PokeData/SecondaryStats.csv')
df2.tail()

number                  name    type1  type2  total   hp  attack  \
1067     896             Glastrier      Ice    NaN    580  100     145   
1068     897             Spectrier    Ghost    NaN    580  100      65   
1069     898               Calyrex  Psychic  Grass    500  100      80   
1070     898     Ice Rider Calyrex  Psychic    Ice    680  100     165   
1071     898  Shadow Rider Calyrex  Psychic  Ghost    680  100      85   

      defense  sp_attack  sp_defense  speed  generation  legendary  
1067      130         65         110     30           8       True  
1068       60        145          80    130           8       True  
1069       80         80          80     80           8       True  
1070      150         85         130     50           8       True  
1071       80        165         100    150           8       True

In [140]:
# connecting to database and creating SQL tables

cnn = sqlite3.connect(':memory:')

df.to_sql('main_stats', cnn)
df2.to_sql('secondary_stats', cnn)

1072

In [141]:
# join tables 

query = '''
SELECT main_stats.ID, main_stats.Name, main_stats.HP, main_stats.Attack, main_stats.Defense, main_stats.SpAtk, main_stats.SpDef, main_stats.Speed,
main_stats.Type1, main_stats.Type2, secondary_stats.generation, secondary_stats.legendary
FROM main_stats
LEFT JOIN secondary_stats
ON main_stats.ID = secondary_stats.number;
'''

result = pd.read_sql_query(query, con=cnn)
result.tail()

ID          Name   HP  Attack  Defense  SpAtk  SpDef  Speed     Type1  \
1603  1021   Raging Bolt  125      73       91    137     89     75  Electric   
1604  1022  Iron Boulder   90     120       80     68    108    124      Rock   
1605  1023    Iron Crown   90      72      100    122    108     98     Steel   
1606  1024     Terapagos   90      65       85     65     85     60    Normal   
1607  1025     Pecharunt   88      88      160     88     88     88    Poison   

        Type2  generation  legendary  
1603   Dragon         NaN        NaN  
1604  Psychic         NaN        NaN  
1605  Psychic         NaN        NaN  
1606     None         NaN        NaN  
1607    Ghost         NaN        NaN

The number of entries in both tables is different meaning that the generation and legendary columns in the result dataset contain missing values. Lets fix this.

In [142]:
# update tables to show generation and legendary status for all null values

result.to_sql('result', cnn)

query = '''
UPDATE result
SET legendary = True
WHERE ID IN (1001, 1002, 1003, 1004, 1007, 1008, 1014, 1015, 1016, 1017, 1024);
'''

query2 = '''
UPDATE result
SET legendary = False
WHERE legendary IS NULL
'''

# the first pokémon in generation 9 has an ID of 899
query3 = '''
UPDATE result
SET generation = 9
WHERE ID > 898
'''

cnn.execute(query)
cnn.execute(query2)
cnn.execute(query3)
cnn.commit()

test_query = '''
SELECT *
FROM result;
'''

filtered_result = pd.read_sql_query(test_query, con=cnn)
filtered_result.tail()

index    ID          Name   HP  Attack  Defense  SpAtk  SpDef  Speed  \
1603   1603  1021   Raging Bolt  125      73       91    137     89     75   
1604   1604  1022  Iron Boulder   90     120       80     68    108    124   
1605   1605  1023    Iron Crown   90      72      100    122    108     98   
1606   1606  1024     Terapagos   90      65       85     65     85     60   
1607   1607  1025     Pecharunt   88      88      160     88     88     88   

         Type1    Type2  generation  legendary  
1603  Electric   Dragon         9.0        0.0  
1604      Rock  Psychic         9.0        0.0  
1605     Steel  Psychic         9.0        0.0  
1606    Normal     None         9.0        1.0  
1607    Poison    Ghost         9.0        0.0

In [146]:
# deleting duplicate values

delete = '''
DELETE FROM result
WHERE rowid NOT IN (
    SELECT MIN(rowid)
    FROM result
    GROUP BY ID
);
'''

cnn.execute(delete)
cnn.commit()

test_delete = '''
SELECT *
FROM result;
'''

pokemon.to_sql('pokemon', cnn)
pokemon = pd.read_sql_query(test_delete, con=cnn)
pokemon.tail()

index    ID          Name   HP  Attack  Defense  SpAtk  SpDef  Speed  \
1020   1603  1021   Raging Bolt  125      73       91    137     89     75   
1021   1604  1022  Iron Boulder   90     120       80     68    108    124   
1022   1605  1023    Iron Crown   90      72      100    122    108     98   
1023   1606  1024     Terapagos   90      65       85     65     85     60   
1024   1607  1025     Pecharunt   88      88      160     88     88     88   

         Type1    Type2  generation  legendary  
1020  Electric   Dragon         9.0        0.0  
1021      Rock  Psychic         9.0        0.0  
1022     Steel  Psychic         9.0        0.0  
1023    Normal     None         9.0        1.0  
1024    Poison    Ghost         9.0        0.0

We can see now that the number of rows is 1025 which matches the number of pokémon that exist. There are no null values and all the columns have information that we need. The data is now cleaned and ready for us to work with.

## Intro: Querying the Data

In [148]:
# number of pokemon in each generation

query = '''
SELECT COUNT(ID) as pokemon_count, generation
FROM pokemon
GROUP BY generation;
'''

cnn.execute(query)
cnn.commit()

result = pd.read_sql_query(query, con=cnn)
print(result)

   pokemon_count  generation
0              2         0.0
1            151         1.0
2            100         2.0
3            135         3.0
4            107         4.0
5            156         5.0
6             72         6.0
7             86         7.0
8             89         8.0
9            127         9.0


We have found some errors in the data. 2 pokemon are classified as generation 0 which doesn't exist. Also, generations 8 and 9 do not have the correct count. Lets fix this.

In [149]:
# finding the error

query = '''
SELECT ID, Name
FROM pokemon
WHERE generation = 0;
'''

cnn.execute(query)
cnn.commit()

result = pd.read_sql_query(query, con=cnn)
result.head()

ID      Name
0  808    Meltan
1  809  Melmetal

In [150]:
# fixing the problem

query = '''
UPDATE pokemon
SET generation = 7
WHERE ID IN (808, 809)
'''

cnn.execute(query)
cnn.commit()

test_query = '''
SELECT ID, Name, generation
FROM pokemon
WHERE ID IN (808, 809)
'''

filtered_result = pd.read_sql_query(test_query, con=cnn)
filtered_result.tail()

ID      Name  generation
0  808    Meltan         7.0
1  809  Melmetal         7.0

In [153]:
# finding the error

query = '''
SELECT ID, Name, generation
FROM pokemon
WHERE generation = 9
LIMIT 10;
'''

cnn.execute(query)
cnn.commit()

result = pd.read_sql_query(query, con=cnn)
print(result)

    ID                      Name  generation
0  899                   Wyrdeer         9.0
1  900                   Kleavor         9.0
2  901                  Ursaluna         9.0
3  902          Basculegion Male         9.0
4  903                  Sneasler         9.0
5  904                  Overqwil         9.0
6  905  Enamorus Incarnate Forme         9.0
7  906                Sprigatito         9.0
8  907                 Floragato         9.0
9  908               Meowscarada         9.0


In [156]:
# fixing the error
# the pokemon with id ranging from 899 to 905 are classified as generation 9 rather than generation 8

query = '''
UPDATE pokemon
SET generation = 8
WHERE ID >= 899 AND ID <= 905
'''

cnn.execute(query)
cnn.commit()

test_query = '''
SELECT ID, Name, generation
FROM pokemon
WHERE ID >= 899 AND ID <= 905
'''

filtered_result = pd.read_sql_query(test_query, con=cnn)
print(filtered_result)

    ID                      Name  generation
0  899                   Wyrdeer         8.0
1  900                   Kleavor         8.0
2  901                  Ursaluna         8.0
3  902          Basculegion Male         8.0
4  903                  Sneasler         8.0
5  904                  Overqwil         8.0
6  905  Enamorus Incarnate Forme         8.0


In [157]:
# rerunning the query to find count of all pokemon per generation

query = '''
SELECT COUNT(ID) as pokemon_count, generation
FROM pokemon
GROUP BY generation;
'''

cnn.execute(query)
cnn.commit()

result = pd.read_sql_query(query, con=cnn)
print(result)

   pokemon_count  generation
0            151         1.0
1            100         2.0
2            135         3.0
3            107         4.0
4            156         5.0
5             72         6.0
6             88         7.0
7             96         8.0
8            120         9.0
